### PR TITLE
textarea[], field[]: Unify function, fix wrong fallback text

### LINKED ISSUE #1978: Redundant words in book interface
Seems like a recent change:
speg9er

### PR DESCRIPTION
**Pull request by SmallJoker**
Fixes https://github.com/minetest/minetest_game/issues/1978

Corrects the compatibility code after #6510 and the suboptimal fix 6b23cab.
**Instructions:**
Use this test mod. If it shows as expected, it's all ok.
```Lua
core.register_on_joinplayer(function(player)
	core.after(2, function(player_name)
	core.show_formspec(player_name, "lab:testy",
		"size[9,7]"..
		-- Book edit
		"field[0.5,1;3.5,0;title;Title:;" ..
			core.formspec_escape("Taking the hobbits to Isengard") .. "]" ..
		"textarea[0.5,1.5;3.5,4;text;Contents:;" ..
			core.formspec_escape("The hobbits the hobbits\nto Isengard") .. "]" ..
		-- Empty book
		"field[4.5,1;3.5,0;title2;Title:;]" ..
		"textarea[0.5,5.5;3.5,4;text2;Contents:;]" ..
		-- Server description (pre e6e5fa3, 0.4.16 stable)
		"textarea[4,2.3;4.23,2.9;;" ..
[[Blinded by the light
Revved up like a deuce
Another runner in the night
Blinded by the light

Madman drummer bummers
Indians in the summer
With a teenage diplomat
In the dumps with the mumps
As the adolescent pumps]] .. ";]" ..
		-- Readonly textarea with title
		"textarea[4.6,5.85;3.7,1.5;;This is a fancy title" ..
		";Never gonna give you up, never gonna let you down]"
	)
	end, player:get_player_name())
end)
```
**EDIT**
As of 180121 this PR also fixes editing of read-only text fields. (thanks to @red-001 for reporting)

### CODE REVIEW

**Comment by kilbith** [comment]
Any chance that https://github.com/minetest/minetest/pull/5361 could get in by this occasion?

**Comment by SmallJoker** [comment]
@kilbith I'd like to but that's out of scope of this PR, which already contains re-structured code and a bugfix.

**Review by numberZero — COMMENTED** [review]
Works fine.

File: src/gui/guiFormSpecMenu.cpp [hunk 1]
```diff
@@ -982,6 +982,70 @@ void GUIFormSpecMenu::parsePwdField(parserData* data, const std::string &element
...
+#endif
+		if (is_multiline)
+			e = new GUIEditBoxWithScrollBar(spec.fdefault.c_str(), true,
+				Environment, this, spec.fid, rect, is_editable, true);
+		else if (is_editable)
+			e = Environment->addEditBox(spec.fdefault.c_str(), rect, true,
+				this, spec.fid);
+	}
+
+	if (e != nullptr) {
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp]: `Environment->addEditBox` may return `NULL`

File: src/gui/guiFormSpecMenu.cpp [hunk 2]
```diff
@@ -982,6 +982,70 @@ void GUIFormSpecMenu::parsePwdField(parserData* data, const std::string &element
...
+		spec.send = true;
+
+	// Multiline textareas: swap default and label for backwards compat
+	if (!is_editable && is_multiline &&
+			spec.fdefault.empty() && !spec.flabel.empty())
+		spec.flabel.swap(spec.fdefault);
+
+	gui::IGUIEditBox *e = nullptr;
+#if USE_FREETYPE && IRRLICHT_VERSION_MAJOR == 1 && IRRLICHT_VERSION_MINOR < 9
+	if (g_settings->getBool("freetype")) {
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp]: This is a piece of the old code. What cleaner way do you have in mind?

File: src/gui/guiFormSpecMenu.cpp [hunk 3]
```diff
@@ -982,6 +982,70 @@ void GUIFormSpecMenu::parsePwdField(parserData* data, const std::string &element
...
+		e = (gui::IGUIEditBox *) new gui::intlGUIEditBox(spec.fdefault.c_str(),
+			true, Environment, this, spec.fid, rect, is_editable, true);
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp]: Not intentional. Will change to `is_multiline`, also switching to the implicit cast. Thanks for pointing this out.

File: src/gui/guiFormSpecMenu.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (numberZero) [reply src/gui/guiFormSpecMenu.cpp]: Exception should probably be thrown in that case; that would be consistent with `new` throwing on failure.

File: src/gui/guiFormSpecMenu.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (numberZero) [reply src/gui/guiFormSpecMenu.cpp]: Something like
```c++
constexpr bool use_intl_edit_box = USE_FREETYPE && IRRLICHT_VERSION_MAJOR == 1 && IRRLICHT_VERSION_MINOR < 9;
...
if (use_intl_edit_box && g_settings->getBool("freetype"))
```
The include should not be guarded by that `#if` in this case, of course.
Unsure will my example compile reliably in all cases, but anyway, the current intermix of `if` and `#if` is too awful to be kept.

**Comment by SmallJoker** [comment]
@numberZero Thanks for reviewing. I adapted your suggestions.

**Comment by numberZero** [comment]
> The include should not be guarded by that #if in this case, of course.

https://github.com/SmallJoker/minetest/blob/9f5b7080598fa6fd226b981d212ed2bfd60a7fb5/src/gui/guiFormSpecMenu.cpp#L59
Otherwise, it won’t compile without FreeType. I don’t know why was the #include guarded, that file obviously doesn’t depend on FreeType. The implementation doesn’t seem to depend on it as well. The fact that it might be useless without FreeType doesn’t mean it should be disabled that hardly.

**Comment by SmallJoker** [comment]
Ready for reviewing. Rebased.

**Comment by SmallJoker** [comment]
Rebased again, added testing instructions.

File: src/gui/guiFormSpecMenu.cpp [hunk 4]
```diff
@@ -1007,6 +1002,71 @@ void GUIFormSpecMenu::parsePwdField(parserData* data, const std::string &element
...
+		gui::StaticText::add(Environment, spec.flabel.c_str(), rect, false, true,
+			this, spec.fid);
+		return;
+	}
+
+	if (is_editable)
+		spec.send = true;
+
+	// Multiline textareas: swap default and label for backwards compat
+	if (!is_editable && is_multiline &&
```
Comment (nerzhul) [inline src/gui/guiFormSpecMenu.cpp]: else if  (and remove !is_editable)

**Comment by Fixer-007** [comment]
Time to get moving?

**Review by nerzhul — APPROVED** [review]

**Review by rubenwardy — APPROVED** [review]

**Comment by rubenwardy** [comment]
just testing now

**Pull request merged**